pulpcore/cli/rpm/content.py

```diff
@@ -52,74 +65,215 @@ def _sha256_artifact_callback(
+        allowed_with_contexts=(PulpRpmPackageContext,),
+    ),
+    pulp_option(
+        "--sha256",
+        callback=_sha256_callback,
+        expose_value=False,
+        allowed_with_contexts=(
+            PulpRpmPackageContext,
+            PulpRpmModulemdDefaultsContext,
+            PulpRpmModulemdContext,
+        ),
+    ),
 ]
+# create works for modulemd_default/modulemd/package only
+# create assumes "there exists an Artifact..."
 create_options = [
-    click.option("--relative-path", required=True),
-    click.option(
+    pulp_option(
+        "--relative-path",
+        required=True,
+        allowed_with_contexts=(
+            PulpRpmPackageContext,
+            PulpRpmModulemdDefaultsContext,
+            PulpRpmModulemdContext,
+        ),
+    ),
+    pulp_option(
         "--sha256",
         "artifact",
         required=True,
         help=_("Digest of the artifact to use"),
         callback=_sha256_artifact_callback,
+        allowed_with_contexts=(
+            PulpRpmPackageContext,
+            PulpRpmModulemdDefaultsContext,
+            PulpRpmModulemdContext,
+        ),
     ),
 ]
 
-
 content.add_command(list_command(decorators=list_options))
 content.add_command(show_command(decorators=lookup_options))
-content.add_command(create_command(decorators=create_options))
+content.add_command(
+    create_command(
+        decorators=create_options,
+        allowed_with_contexts=(
+            PulpRpmPackageContext,
+            PulpRpmModulemdDefaultsContext,
+            PulpRpmModulemdContext,
+        ),
+    )
+)
 
 
-@content.command()
-@click.option("--relative-path", required=True)
-@click.option("--file", type=click.File("rb"), required=True)
+# upload works for advisory/modulemd_default/modulemd/package
+# upload takes a file-argument and uploads to pulp, creates the artifact,
+# then creates the entity from it
+@content.command(
+    allowed_with_contexts=(
+        PulpRpmAdvisoryContext,
+        PulpRpmPackageContext,
+        PulpRpmModulemdDefaultsContext,
+        PulpRpmModulemdContext,
+    )
+)
 @chunk_size_option
-@pass_entity_context
+@pulp_option(
+    "--relative-path",
+    required=True,
+    help=_("Relative path within a distribution of the entity"),
+    allowed_with_contexts=(
+        PulpRpmPackageContext,
+        PulpRpmModulemdDefaultsContext,
+        PulpRpmModulemdContext,
+    ),
+)
+@pulp_option(
+    "--file",
+    type=click.File("rb"),
+    required=True,
+    help=_("An RPM binary"),
+    allowed_with_contexts=(PulpRpmPackageContext,),
+)
+@pulp_option(
+    "--file",
+    type=click.File("rb"),
+    required=True,
+    help=_("A modulemd/defaults JSON file"),
+    allowed_with_contexts=(PulpRpmModulemdDefaultsContext, PulpRpmModulemdContext),
+)
+@pulp_option(
+    "--file",
+    type=click.File("rb"),
+    required=True,
+    help=_("An advisory JSON file"),
+    allowed_with_contexts=(PulpRpmAdvisoryContext,),
+)
 @pass_pulp_context
+@pass_entity_context
```

Comment:
I think this is the error, you swapped these around so pulp_ctx is actually the entity_ctx in the function.

Reply:
Yes exactly, order matters here. Each pass decorator consumes the first remaining args-slot from inside out. So they appear in reverse order.

Reply:
Yup, exactly - thanks for the pointer, gentlemen!